### PR TITLE
Download preplanned map

### PR DESCRIPTION
**Pull request by JonLavi**
Hi both,
Now that all the new Offline features are wrapped, I'm submitting this PR for `Download Preplanned Map`.

Looking forward to seeing what you think.

### CODE REVIEW

**Comment by Rachael-E** [comment]
@JonLavi thanks for the chat yesterday about the sample, just to document what we discussed (sorry I didn't get round to documenting this yesterday):

1. This sample would benefit from being able to see the outlines of the individual preplanned areas
2. When all 4 areas are downloaded, should we expect to see them all on the map instead of just one (I think so) 
3. When downloading multiple of the same area, that file cannot be deleted (and results in multiple errors) because it looks like the geodatabase is still in use and the temp folder isn't deleted
4. Temp folder isn't deleting on exiting the app
5. The thumbnail image for Country Commons shows the wrong area

File: map/download-preplanned-map/README.metadata.json
Comment (Rachael-E) [inline map/download-preplanned-map/README.metadata.json]: this is written twice

File: map/download-preplanned-map/README.metadata.json
Comment (Rachael-E) [inline map/download-preplanned-map/README.metadata.json]: I think this should be 
"/java/latest/sample-code/download-preplanned-map.htm",

File: map/download-preplanned-map/README.md
Comment (Rachael-E) [inline map/download-preplanned-map/README.md]: We should make some button name clarifications here, and update the button names to reflect what's in the sample:

> Select a map area, then click "Download Preplanned Area" to take it offline. Click "Delete All Downloaded Areas" to remove any downloaded map areas.

Comment (Rachael-E) [reply map/download-preplanned-map/README.md]: We also need to update this section for explaining the "Show Web Map button"

File: map/download-preplanned-map/README.md
Comment (Rachael-E) [inline map/download-preplanned-map/README.md]: > Start the job. Once it has completed, get the `DownloadPreplannedOfflineMapResult`.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: I'd move the instantiation of the graphics overlay and the renderer to line 117 (after `mapView.setMap(webMap)` ) to keep code order and format consistent across our samples repo and to keep the logic in order i.e. create Map, create MapView, set map to mapView, create GraphicsOverlay, set graphicsOverlay to mapView.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: Line 92 specifies the preplanned map area as an area of interest, but I think it's more helpful to continue referring to them as preplanned map areas rather than areas of interest. "Preplanned Map Area" is specific and clear to the user what is being referred to, "Area of Interest" is a little more ambiguous.

Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: create a symbol to mark the areas of interest -> create a red outline to mark the preplanned map areas

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: -> create a simple renderer for the preplanned map areas with the simple line symbol

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: "make a list to store opened mobile map packages"

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: create a listener that changed the viewpoint -> create a listener that *changes* the viewpoint

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: can we change this variable to be preplannedMapAreaOutline, I find area of interest not specific enough.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: let's make these method names a bit clearer: 
`handleDownloadPreplannedAreaButtonClicked`

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: "display the map from the mobile map package"

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: full stop at the end of this sentence

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: We should add some more info in here. It's not only downloading the selected area, but it's then also checking to see if it's been downloaded, and then showing the result. e.g.

"Downloads and shows the selected preplanned map area"

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: I think we should move this block handling if it's already downloaded to the end of the method: it's better to show upfront how to go about downloading the chosen preplanned area. e.g. it can go on line 290 before the final `else`. Can also remove the `return` at the end of this block when moved there

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java:209]: Matter of preference perhaps, but for readability we can remove some of the new lines here between the set up code for this method (so that the user can get to the download preplanned code sooner). e.g. I'd remove the new spaces between lines 197 and 207.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: we can change this method name to: 

`handleShowWebMapButtonClicked`

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: map vew -> map view

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: `handleDeleteOfflineAreasButtonClicked`

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: is it necessary to cancel the job here if we're just deleting the existing offline areas? We defo need the close the geodatabase, I'm just not convinced we need to cancel the job too.  It's my understanding that cancelling would be if you wanted to stop the job downloading mid download.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: Stops the animation?

File: map/download-preplanned-map/src/main/resources/download_preplanned_map.fxml
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/resources/download_preplanned_map.fxml]: can remove empty line here

File: map/download-preplanned-map/src/main/resources/style.css
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/resources/style.css:9]: I'm not too sure what the approach for .css files have been with splitting up the samples: do we share a single css file now or does each sample have it's own css? I suspect the latter, in which case this .css file contains unused elements such as slider and checkbox.

**Review by Rachael-E — CHANGES_REQUESTED** [review]
@JonLavi thanks very much for all your hard work going into this. The UI is looking good now and the sample flows well. Most of my comments are clarifications/renaming/refactoring suggestions, and some changes are needed in the .css file and in the metadata file. 

A few other things: 
- The .png should be updated to reflect the new design
- A file called 'q' with 16193 lines has crept in here somehow, looks like a commit history document!

File: map/download-preplanned-map/README.metadata.json
Comment (JonLavi) [reply map/download-preplanned-map/README.metadata.json]: Good spot, the second line was supposed to read '*Launcher.java'

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (JonLavi) [reply map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: How about the slightly more specific 'preplanned map area extent'? 
The API does refer to it as area of interest though (`PreplannedMapArea.getAreaofInterest()`), and that's essentially what we're drawing a line around.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (JonLavi) [reply map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: I find it makes sense to check whether the area has already been downloaded at the very beginning though, to avoid downloading it again. We could flip it around and check whether there _arent_ any files:
```
if (!Files.exists(Paths.get(path))) {
 // download the map
} else {
 // open the map
}
```
Although I'd have to think about the implications of this.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (JonLavi) [reply map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: I agree that these names are more specific, but since the methods get called at some places not as a result of a button click, I don't think it would be appropriate to add the buttons to the method name. e.g. `showWebMap()` and `closeAllGeoDatabases()` get called within `deleteOfflineAreas()`. The last one also gets called in `terminate()`.
That's semantics though, so happy to rename if you still think that's appropriate.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (JonLavi) [reply map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: Snuck in from a Scene sample, I think! >.<

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (JonLavi) [reply map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: I took over this block from the .NET sample. Agree that with the way it's set up, it's not possible to trigger this method because the button is disabled, and I think it was the same in .NET. Perhaps it's simply advocating good practice, that any potential job should be stopped and cancelled?

Comment (JonLavi) [reply map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: As discussed, removing this as it seems unnecessary

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (JonLavi) [reply map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: With the discussed refactoring this method is now renamed as per your suggestion 👍

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (JonLavi) [reply map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: Turned this statement around as described below. Definitely reads nicer when the block for downloading appears first. Some minor refactoring as well as a result. Let me know what you think.

File: map/download-preplanned-map/README.md
Comment (Rachael-E) [inline map/download-preplanned-map/README.md:15]: If we're updating the README to show clicking on a "Show Online Map" button, then the button text, related methods and descriptions also need to be updated in the code.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java:108]: we should refer to the "webMap" as the "onlineMap" following Lucas's feedback.

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: `showOnlineMapButton.setDisable(false)`

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: enable the 'view online map' button

File: map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java
Comment (Rachael-E) [inline map/download-preplanned-map/src/main/java/com/esri/samples/download_preplanned_map/DownloadPreplannedMapController.java]: this method should be renamed now to `showOnlineMap`

**Review by Rachael-E — COMMENTED** [review]
thanks @JonLavi, it's looking good to me. Just need to update the references of web map to online map, and make sure the .png is up to date with the current design.

**Comment by JonLavi** [comment]
Thanks! The web-map references are all fixed, and holding off on the screenshot until we've made a final decision on the UI

**Comment by Rachael-E** [comment]
@JonLavi thanks, if you could push that commit fixing the webMap references when you get a moment, I can approve and then it's time for a second review.

**Review by Rachael-E — APPROVED** [review]
thanks @JonLavi, this is ready for a second review now

**Review by tschie — COMMENTED** [review]
- button can just say 'Download' instead of 'Download Preplanned Map Areas'.
- download button should be disabled if no list item is selected (use a property binding)
- if you are viewing a downloaded area and you select a different item in the list and then click view online map, the viewpoint should go to the selected item when the online map is shown.

File: map/download-preplanned-map/README.md
Comment (tschie) [inline map/download-preplanned-map/README.md]: I wasn't prompted for credentials.

File: map/download-preplanned-map/README.md
Comment (tschie) [inline map/download-preplanned-map/README.md]: Don't need the (see additional info...) section here.

Comment (tschie) [reply map/download-preplanned-map/README.md]: You could just have this say "Set the update mode of the preplanned map area".

**Comment by tschie** [comment]
Closing this one in favor of the merged PR linked above

**Pull request closed without merging**
